### ISSUE TITLE
Desync on MP with slow connection

### ISSUE THREAD

**Issue by TimVroom** [h8ul-modder/factorio-ribbon-maze#38]
Hi.

Bit hard to explain/debug, but sometimes people with a somewhat slower connection to my server are getting desyncs. From their desync reports i
 could figure out that the rng generator was giving different results. (See [desync-report-2019-03-13_12-49-45.zip](https://github.com/h8ul-modder/factorio-ribbon-maze/files/2963338/desync-report-2019-03-13_12-49-45.zip))

It seems to be in the firstMazeRowMangrove variable where they differ sometimes.
This was tested on a 0.17 install, so this might be the issue (since the PR for 0.17 isn't accepted yet).

We also installed some other mods, but this is the only difference i found in the script.dat that would/could generate these issues.

I've been trying to fix / reproduce it myself, but cannot seem to do that with the lack of knowledge i have.

Anyhow in this case it seems that the 7th iteration of the chunk was giving issues.
The server is a linux headless, the client on windows (all cases), but not all windows clients seem to get this result. Most/

At least it comes from the Cmwc.deriveNew function it seems.
And as far as i can see, only the 7th chunk was giving issues. There are no other mods that hook into the chunk generation, we have stdlib installed, but that doesn't seem to influence it.

And its also that its not immidiatly desyncing, but after a second or 10 - 15

If you need more information, please let me know. 

P.s. this is with the 0.17 PR included.

**Comment by h8ul-modder** [comment]
Would you be able to try running a game with the maze block size at the default one chunk (32x32)?

I have a theory that chunks are being generated in a different order for different players. That shouldn't affect the default block size, because RibbonMaze was designed to make sure that the order of chunk generation doesn't matter -- in fact that's the job of Cmwc.deriveNew(), a new deterministic pseudorandom number generator is derived for each chunk in a strict order, regardless of when the chunk is generated. This was mainly to make sure mapgen seeds have reliable results, but there was an element of paranoia around desyncs of exactly the nature you've raised.

Except that the option to make the block size bigger was added into the mod later, and that relaxed some of the strict logic. Cmwc.deriveNew() is called per "block" not per "chunk" for mangroves and resources -- so there are four chunks in each of the blocks with the settings you've got sharing a single RNG. Therefore, if factorio doesn't emit chunk generation events in the same order on each machine, then the resources and mangroves will differ, and *boom* -- desync.

The nightmare is now real.

If this is the case, and we can prove it, then it is presumably a bug in 0.17; it's hard for mods to be deterministic if the events generated by the core are not. If it's really the responsibility of my mod to be immune to different player's machines emitting chunk generation events in different orders, then it's going to impact memory usage and save size, because I'll need to derive more random number generators to defend against it. But one step at a time.

**Comment by TimVroom** [comment]
Been away for a little while, but even with the 32 chunk size it's giving desyncs.
And tested with only the maze runner mod, so it might seem you are right, that because of the lagg/delay it generates the chunk differently.

And it seems thats only with slower connections.
The desnyc happened as soon as the one on the slower connection generated a new chunk.

Hope you can figure it out and that its not a 0.17 core bug

[desync-report-2019-03-22_22-06-54.zip](https://github.com/h8ul-modder/factorio-ribbon-maze/files/2998336/desync-report-2019-03-22_22-06-54.zip)

**Comment by h8ul-modder** [comment]
Very much appreciating your help figuring this out!

I was anticipating no desyncs on block size 32, but I've figured out that the fish generation is using the wrong RNG, which could result in different chunk generation if the chunks are generated in a different order on two machines regardless of block size. I've released a fix in release 0.7.1.

Would you be able to test again at block size 32? I think if the problem is now solved at block size 32, that really tells us something very interesting.

**Comment by TimVroom** [comment]
I've tried it with 0.7.1
Just update the mod with the old save game. 
Same map, 32 block size and only ribbon maze as mod installed.

I didn't have problems, the tester who had desyncs before immidiatly desyncs (within 18 ticks of joining)

So it didn't seem to help (enough)


```
5853.646 Info ServerMultiplayerManager.cpp:837: updateTick(10478) received stateChanged peerID(3) oldState(TryingToCatchUp) newState(WaitingForCommandToStartSendingTickClosures)
5853.646 Info GameActionHandler.cpp:4019: UpdateTick (10478) processed PlayerJoinGame peerID(3) playerIndex(1) mode(connect)
5853.748 Info ServerMultiplayerManager.cpp:837: updateTick(10484) received stateChanged peerID(3) oldState(WaitingForCommandToStartSendingTickClosures) newState(InGame)
2019-03-26 19:26:22 [JOIN] XXXXX joined the game
5870.931 Info ServerMultiplayerManager.cpp:837: updateTick(11515) received stateChanged peerID(3) oldState(InGame) newState(DesyncedWaitingForMap)
5870.931 Info ServerMultiplayerManager.cpp:917: updateTick(11515) received playerDesynced peer (3)
```

[desync-report-2019-03-26_20-26-51.zip](https://github.com/h8ul-modder/factorio-ribbon-maze/files/3009986/desync-report-2019-03-26_20-26-51.zip)

**Comment by h8ul-modder** [comment]
Did the tester have success connecting to your game with Ribbon Maze in 0.16?

I can't think of any reason remaining, why the firstMazeRowMangroveRng would differ between two systems, unless maybe the chunk is being generated twice on one machine and once on the other. And if that's happening, I think that's a bug in the core.

**Comment by TimVroom** [comment]
He was able to connect without any problem on 0.16, never had a desync.
And in the last one the 20th version of the {Q={...}} was giving errors. Seems different every time. However the first few was in within the first 10 rng numbers, now its in the 20's...

And with this one its the second to last thats different. I would expect more to be different, but it seems it's only one. Might be a core bug though :(

At least my knowledge stops with just minor debugging... where it's actually comming from... no idea. I hope you can figure it out. Maybe the Factorio devs can help you out.

**Comment by TimVroom** [comment]
Have been doing a few extra tests from my mac with connection and slowness:

Have been trying to connect through multiple vpn's throughout the world. But none gave me desyncs. So seems to be a windows only issues.
If you need some additional information please let me know.

**Comment by h8ul-modder** [comment]
This is believed to be fixed in 0.7.4 as I was able to reproduce the issue with a seed of "1" generating different maps on Windows and Linux. Problem was with Lua having different arithmetic behaviour on those platforms. This has been worked around.

Closing unless desync reports are received for new multiplayer maps.
